TrafficCapture/trafficReplayer/src/main/java/org/opensearch/migrations/replay/kafka/KafkaProtobufConsumer.java

```diff
@@ -20,25 +24,108 @@
 import java.time.Duration;
 import java.util.Collections;
 import java.util.List;
-import java.util.Map;
 import java.util.Objects;
+import java.util.Optional;
+import java.util.PriorityQueue;
 import java.util.Properties;
 import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 
+/**
+ * Adapt a Kafka stream into a TrafficCaptureSource.
+ *
+ * Notice that there's a critical gap between how Kafka accepts commits and how the
+ * BlockingTrafficSource throttles calls to Kafka.  The BlockingTrafficSource may
+ * block calls to readNextTrafficStreamChunk() until some time window elapses.  This
+ * could be a very large window in cases where there were long gaps between recorded
+ * requests from the capturing proxy.  For example, if a TrafficStream is read and it
+ * that stream is scheduled to be run one hour later, readNextTrafficStreamChunk()
+ * may not be called for almost an hour.  By design, we're not calling Kafka to pull
+ * any more messages since we know that we don't have work to do for an hour.  Shortly
+ * after the hour of waiting begins, Kakfa will notice that this application is no
+ * longer calling poll and will kick the consumer out of the client group.  Other
+ * consumers may connect, though they'll also be kicked out of the group shortly.
+ * It may be fine to have this pattern of cascaded replayers waiting to send a
+ * request, then all of them sending at about the same time....
+ *
+ * See
+ * <a href="https://kafka.apache.org/21/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#failuredetection">...</a>
+ *
+ * "Basically if you don't call poll at least as frequently as the configured max interval,
+ * then the client will proactively leave the group so that another consumer can take
+ * over its partitions. When this happens, you may see an offset commit failure (as
+ * indicated by a CommitFailedException thrown from a call to commitSync())."
+ *
+ * I believe that this can be mitigated, hopefully fully, by adding a keepAlive/do nothing
```

Comment:
What happens with these recent changes when we have not polled in a while and the consumer leaves the group? We will hit a commit failure and try to process again?

Reply:
We'll fail and forget. Though as other messages are processed, we have the potential in the future to not fail. 
However, notice that there are some patterns where we could NEVER successfully commit.